README.md
[![npm version](https://badge.fury.io/js/@dellacagna%2Fiban.svg)](https://badge.fury.io/js/@dellacagna%2Fiban)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/dellacagna/iban/master/LICENSE)

# IBAN Library
A simple library for validating, formatting and generating fake IBAN numbers.

This library follows the [ISO 13616 IBAN Registry technical specification](https://www.swift.com/standards/data-standards/iban-international-bank-account-number) and supports 107 countries.

## Usage
### NodeJS
```ts
import { IBAN } from '../index';  
let iban = new IBAN('GB82 WEST 1234 5698 7654 32');
if (iban.isValid()) {
    console.log('The IBAN is valid');
}
```
## API
### isValid()
Returns **true** for a valid IBAN or **false** if invalid.

example:
```ts
let iban = new IBAN('GB82 WEST 1234 5698 7654 32');
if (iban.isValid()) {
    console.log('The IBAN is valid');
}
```

### getError()
If the IBAN is invalid, one of the following errors will be returned as a string:
```
INVALID_COUNTRY
INVALID_LENGTH
INVALID_FORMAT
CHECKSUM_FAILED
```


example:
```ts

let iban = new IBAN('GB82 WEST 1234');
if (!iban.isValid()) {
    console.log(!iban.getError()); //prints INVALID_LENGTH
}
```

### getOriginal()
Returns the original unformatted IBAN that was used to instantiate the object.

example:
```ts
let iban = new IBAN('GB82 weST 123456987654   32');
let unformatted = iban.getOriginal()
console.log(unformatted); //prints GB82 weST 123456987654   32
```

### toElectronicFormat()
Returns the IBAN capitalized with all spaces removed.

example:
```ts
let iban = new IBAN('GB82 weST 123456987654   32');
let formatted = iban.toElectronicFormat()
console.log(formatted); //prints GB82WEST12345698765432
```

### toPrintFormat()
Returns the IBAN capitalized with spaces placed as per country format.

example:
```ts
let iban = new IBAN('GB82 weST 123456987654   32');
let formatted = iban.toPrintFormat()
console.log(formatted); //prints GB82 WEST 1234 5698 7654 32
```

### getBBAN()
Returns the BBAN (Basic Bank Account Number) capitalized with all spaces removed.

example:
```ts
let iban = new IBAN('GB82 WeST 1234 5698 7654 32');
let bban = iban.getBBAN()
console.log(bban); //prints WEST12345698765432
```

### getFields()
Returns an object with extracted information from a valid IBAN. Empty keys will not be returned:

```
{
  countryCode: string;
  country: string;
  checkDigits: string;
  accountNumber?: string;
  currencyCode?: string;
  ownerAccountNumber?: string;
  accountNumberPrefix?: string;
  bicBankCode?: string;
  branchCode?: string;
  accountType?: string;
  nationalCheckDigit?: string;
 }
```

example:
```ts
let iban = new IBAN('BA39 1290 0794 0102 8494');
let fields = iban.getFields()
console.log(fields);

/* This will output:
  {
    countryCode: 'BA',
    country: 'Bosnia and Herzegovina',
    checkDigits: '39',
    accountNumber: '94010284',
    bankCode: '129',
    branchCode: '007',
    nationalCheckDigit: '94',
 } 
*/
```
### random()
Generate a random IBAN number and return it as a string in print format. You can optionally pass the country code to specify a random IBAN for that country

example (random country):
```ts
let iban = IBAN.random();
console.log(iban); //example output: GB82 WEST 1234 5698 7654 32
```

example (specified country):
```ts
let iban = IBAN.random('IT');
console.log(iban); //example output: IT60 X054 2811 1010 0000 0123 456
```

## Resources
* [Changelog](https://github.com/dellacagna/iban/blob/master/CHANGELOG.md)
* [License](https://raw.githubusercontent.com/dellacagna/iban/master/LICENSE)


CHANGELOG.md
Changelog
=========

### 1.1.0

* Release Jan 24, 2023
  * Added IBAN generator

### 1.0.0

* Release Jan 23, 2023
    * Added a country specification list for countries supporting IBAN to validate against
    * It is possible to format IBANs into electronic and print formats
    * Extract BBAN
    * Extract data from Iban such as BIC

### 0.0.1

* Release Jan 16, 2023
    * Initial Release supporting only checksum validation.


src/lib/specifications.ts
export interface IbanSpecification {
  code: string;
  country: string;
  length: number;
  formatRules: string[];
  spacingRules: number[];
  fields: string;
  hasNationalCheck: boolean;
}

export const specifications: IbanSpecification[] = [
  {
    code: 'AD',
    country: 'Andorra',
    length: 24,
    formatRules: ['\\d{8}', '[A-Z\\d]{12}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBSSSSCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'AE',
    country: 'United Arab Emirates',
    length: 23,
    formatRules: ['\\d{3}', '\\d{16}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'AL',
    country: 'Albania',
    length: 28,
    formatRules: ['\\d{8}', '[A-Z\\d]{16}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSSXCCCCCCCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'AO',
    country: 'Angola',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'AT',
    country: 'Austria',
    length: 20,
    formatRules: ['\\d{16}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBBCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'AZ',
    country: 'Azerbaijan',
    length: 28,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'BA',
    country: 'Bosnia and Herzegovina',
    length: 20,
    formatRules: ['\\d{16}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBSSSCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'BE',
    country: 'Belgium',
    length: 16,
    formatRules: ['\\d{12}'],
    spacingRules: [4, 9, 14],
    fields: 'BBBCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'BF',
    country: 'Burkina Faso',
    length: 28,
    formatRules: ['[A-Z\\d]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'BG',
    country: 'Bulgaria',
    length: 22,
    formatRules: ['[A-Z]{4}', '\\d{6}', '[A-Z\\d]{8}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBSSSSTTCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'BH',
    country: 'Bahrain',
    length: 22,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{14}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'BI',
    country: 'Burundi',
    length: 16,
    formatRules: ['\\d{12}'],
    spacingRules: [4, 9, 14],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'BJ',
    country: 'Benin',
    length: 28,
    formatRules: ['[A-Z\\d]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'BR',
    country: 'Brazil',
    length: 29,
    formatRules: ['\\d{23}', '[A-Z]{1}', '[A-Z\\d]{1}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBBBBBSSSSSCCCCCCCCCCTN',
    hasNationalCheck: false,
  },
  {
    code: 'BY',
    country: 'Belarus',
    length: 28,
    formatRules: ['[A-Z\\d]{4}', '\\d{4}', '[A-Z\\d]{16}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBAAAACCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'CF',
    country: 'Central African Republic',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'CG',
    country: 'Congo, Republic of the',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'CH',
    country: 'Switzerland',
    length: 21,
    formatRules: ['\\d{5}', '[A-Z\\d]{12}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBBCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'CI',
    country: "Côte d'Ivoire",
    length: 28,
    formatRules: ['[A-Z]{1}', '\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'CM',
    country: 'Cameroon',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'CR',
    country: 'Costa Rica',
    length: 22,
    formatRules: ['\\d{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: '0BBBCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'CV',
    country: 'Cabo Verde',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'CY',
    country: 'Cyprus',
    length: 28,
    formatRules: ['\\d{8}', '[A-Z\\d]{16}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSSSCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'CZ',
    country: 'Czech Republic',
    length: 24,
    formatRules: ['\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBPPPPSSCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'DE',
    country: 'Germany',
    length: 22,
    formatRules: ['\\d{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBBBBBCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'DJ',
    country: 'Djibouti',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'DK',
    country: 'Denmark',
    length: 18,
    formatRules: ['\\d{14}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBCCCCCCCCCX',
    hasNationalCheck: false,
  },
  {
    code: 'DO',
    country: 'Dominican Republic',
    length: 28,
    formatRules: ['[A-Z]{4}', '\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'DZ',
    country: 'Algeria',
    length: 26,
    formatRules: ['\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'EE',
    country: 'Estonia',
    length: 20,
    formatRules: ['\\d{16}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBSSCCCCCCCCCCCX',
    hasNationalCheck: true,
  },
  {
    code: 'EG',
    country: 'Egypt',
    length: 29,
    formatRules: ['\\d{25}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBSSSSCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'ES',
    country: 'Spain',
    length: 24,
    formatRules: ['\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBSSSSXXCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'FI',
    country: 'Finland',
    length: 18,
    formatRules: ['\\d{14}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBBBCCCCCCCX',
    hasNationalCheck: true,
  },
  {
    code: 'FO',
    country: 'Faroe Islands',
    length: 18,
    formatRules: ['\\d{14}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBCCCCCCCCCX',
    hasNationalCheck: false,
  },
  {
    code: 'FR',
    country: 'France',
    length: 27,
    formatRules: ['\\d{10}', '[A-Z\\d]{11}', '\\d{2}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBBSSSSSCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'GA',
    country: 'Gabon',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'GB',
    country: 'United Kingdom',
    length: 22,
    formatRules: ['[A-Z]{4}', '\\d{14}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBSSSSSSCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'GE',
    country: 'Georgia',
    length: 22,
    formatRules: ['[A-Z\\d]{2}', '\\d{16}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'GI',
    country: 'Gibraltar',
    length: 23,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{15}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'GL',
    country: 'Greenland',
    length: 18,
    formatRules: ['\\d{14}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBCCCCCCCCCX',
    hasNationalCheck: false,
  },
  {
    code: 'GQ',
    country: 'Equatorial Guinea',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'GR',
    country: 'Greece',
    length: 27,
    formatRules: ['\\d{7}', '[A-Z\\d]{16}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSSCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'GT',
    country: 'Guatemala',
    length: 28,
    formatRules: ['[A-Z\\d]{4}', '[A-Z\\d]{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBMMTTCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'GW',
    country: 'Guinea-Bissau',
    length: 25,
    formatRules: ['[A-Z\\d]{2}', '\\d{19}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'HN',
    country: 'Honduras',
    length: 28,
    formatRules: ['[A-Z]{4}', '\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'HR',
    country: 'Croatia',
    length: 21,
    formatRules: ['\\d{17}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBBBBCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'HU',
    country: 'Hungary',
    length: 28,
    formatRules: ['\\d{24}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSSXCCCCCCCCCCCCCCCX',
    hasNationalCheck: true,
  },
  {
    code: 'IE',
    country: 'Ireland',
    length: 22,
    formatRules: ['[A-Z\\d]{4}', '\\d{14}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'QQQQBBBBBBCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'IL',
    country: 'Israel',
    length: 23,
    formatRules: ['\\d{19}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBSSSCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'IQ',
    country: 'Iraq',
    length: 23,
    formatRules: ['[A-Z]{4}', '\\d{15}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBSSSCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'IR',
    country: 'Iran',
    length: 26,
    formatRules: ['\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'IS',
    country: 'Iceland',
    length: 26,
    formatRules: ['\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBSSTTCCCCCCIIIIIIIIII',
    hasNationalCheck: true,
  },
  {
    code: 'IT',
    country: 'Italy',
    length: 27,
    formatRules: ['[A-Z]{1}', '\\d{10}', '[A-Z\\d]{12}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'XBBBBBSSSSSCCCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'JO',
    country: 'Jordan',
    length: 30,
    formatRules: ['[A-Z]{4}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBSSSSCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'KM',
    country: 'Comoros',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'KW',
    country: 'Kuwait',
    length: 30,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'KZ',
    country: 'Kazakhstan',
    length: 20,
    formatRules: ['\\d{3}', '[A-Z\\d]{13}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LB',
    country: 'Lebanon',
    length: 28,
    formatRules: ['\\d{4}', '[A-Z\\d]{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LC',
    country: 'Saint Lucia',
    length: 32,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{24}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LI',
    country: 'Liechtenstein',
    length: 21,
    formatRules: ['\\d{5}', '[A-Z\\d]{12}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBBCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LT',
    country: 'Lithuania',
    length: 20,
    formatRules: ['\\d{16}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBBCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LU',
    country: 'Luxembourg',
    length: 20,
    formatRules: ['\\d{3}', '[A-Z\\d]{13}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LV',
    country: 'Latvia',
    length: 21,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{13}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'LY',
    country: 'Libya',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'MA',
    country: 'Morocco',
    length: 28,
    formatRules: ['\\d{24}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'MC',
    country: 'Monaco',
    length: 27,
    formatRules: ['\\d{10}', '[A-Z\\d]{11}', '\\d{2}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBBSSSSSCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'MD',
    country: 'Moldova',
    length: 24,
    formatRules: ['[A-Z\\d]{2}', '[A-Z\\d]{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'ME',
    country: 'Montenegro',
    length: 22,
    formatRules: ['\\d{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'MG',
    country: 'Madagascar',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'MK',
    country: 'North Macedonia',
    length: 19,
    formatRules: ['\\d{3}', '[A-Z\\d]{10}', '\\d{2}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'ML',
    country: 'Mali',
    length: 28,
    formatRules: ['[A-Z\\d]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'MR',
    country: 'Mauritania',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBBSSSSSCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'MT',
    country: 'Malta',
    length: 31,
    formatRules: ['[A-Z]{4}', '\\d{5}', '[A-Z\\d]{18}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBSSSSSCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'MU',
    country: 'Mauritius',
    length: 30,
    formatRules: ['[A-Z]{4}', '\\d{19}', '[A-Z]{3}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBBBSSCCCCCCCCCCCC000MMM',
    hasNationalCheck: false,
  },
  {
    code: 'MZ',
    country: 'Mozambique',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'NE',
    country: 'Niger',
    length: 28,
    formatRules: ['[A-Z]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'NI',
    country: 'Nicaragua',
    length: 32,
    formatRules: ['[A-Z]{4}', '\\d{24}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'NL',
    country: 'Netherlands',
    length: 18,
    formatRules: ['[A-Z]{4}', '\\d{10}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'NO',
    country: 'Norway',
    length: 15,
    formatRules: ['\\d{11}'],
    spacingRules: [4, 9, 14],
    fields: 'BBBBCCCCCCX',
    hasNationalCheck: true,
  },
  {
    code: 'PK',
    country: 'Pakistan',
    length: 24,
    formatRules: ['[A-Z\\d]{4}', '\\d{16}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'PL',
    country: 'Poland',
    length: 28,
    formatRules: ['\\d{24}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBSSSSXCCCCCCCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'PS',
    country: 'Palestinian territories',
    length: 29,
    formatRules: ['[A-Z\\d]{4}', '\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'PT',
    country: 'Portugal',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBSSSSCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'QA',
    country: 'Qatar',
    length: 29,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'RO',
    country: 'Romania',
    length: 24,
    formatRules: ['[A-Z]{4}', '[A-Z\\d]{16}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'RS',
    country: 'Serbia',
    length: 22,
    formatRules: ['\\d{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'RU',
    country: 'Russia',
    length: 33,
    formatRules: ['\\d{14}', '[A-Z\\d]{15}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34, 39],
    fields: 'BBBBBBBBBSSSSSCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'SA',
    country: 'Saudi Arabia',
    length: 24,
    formatRules: ['\\d{2}', '[A-Z\\d]{18}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'SC',
    country: 'Seychelles',
    length: 31,
    formatRules: ['[A-Z]{4}', '\\d{20}', '[A-Z]{3}'],
    spacingRules: [4, 9, 12, 15, 20, 25, 30, 35],
    fields: 'BBBBBBSSCCCCCCCCCCCCCCCCMMM',
    hasNationalCheck: false,
  },
  {
    code: 'SD',
    country: 'Sudan',
    length: 18,
    formatRules: ['\\d{14}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'SE',
    country: 'Sweden',
    length: 24,
    formatRules: ['\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCCCCX',
    hasNationalCheck: false,
  },
  {
    code: 'SI',
    country: 'Slovenia',
    length: 19,
    formatRules: ['\\d{15}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBSSSCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'SK',
    country: 'Slovakia',
    length: 24,
    formatRules: ['\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBPPPPSSCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'SM',
    country: 'San Marino',
    length: 27,
    formatRules: ['[A-Z]{1}', '\\d{10}', '[A-Z\\d]{12}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'XBBBBBSSSSSCCCCCCCCCCCC',
    hasNationalCheck: true,
  },
  {
    code: 'SN',
    country: 'Senegal',
    length: 28,
    formatRules: ['[A-Z]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'ST',
    country: 'São Tomé and Príncipe',
    length: 25,
    formatRules: ['\\d{21}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBSSSSCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'SV',
    country: 'El Salvador',
    length: 28,
    formatRules: ['[A-Z]{4}', '\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBCCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'TD',
    country: 'Chad',
    length: 27,
    formatRules: ['\\d{23}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'TG',
    country: 'Togo',
    length: 28,
    formatRules: ['[A-Z]{2}', '\\d{22}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: '',
    hasNationalCheck: false,
  },
  {
    code: 'TL',
    country: 'East Timor',
    length: 23,
    formatRules: ['\\d{19}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'TN',
    country: 'Tunisia',
    length: 24,
    formatRules: ['\\d{20}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBSSSCCCCCCCCCCCCCXX',
    hasNationalCheck: true,
  },
  {
    code: 'TR',
    country: 'Turkey',
    length: 26,
    formatRules: ['\\d{5}', '[A-Z\\d]{17}'],
    spacingRules: [4, 9, 14, 19, 24, 29],
    fields: 'BBBBB0CCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'UA',
    country: 'Ukraine',
    length: 29,
    formatRules: ['\\d{6}', '[A-Z\\d]{19}'],
    spacingRules: [4, 9, 14, 19, 24, 29, 34],
    fields: 'BBBBBBCCCCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'VA',
    country: 'Vatican City',
    length: 22,
    formatRules: ['\\d{3}', '\\d{15}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'VG',
    country: 'Virgin Islands, British',
    length: 24,
    formatRules: ['[A-Z\\d]{4}', '\\d{16}'],
    spacingRules: [4, 9, 14, 19, 24],
    fields: 'BBBBCCCCCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
  {
    code: 'XK',
    country: 'Kosovo',
    length: 20,
    formatRules: ['\\d{4}', '\\d{10}', '\\d{2}'],
    spacingRules: [4, 9, 14, 19],
    fields: 'BBBBCCCCCCCCCCCC',
    hasNationalCheck: false,
  },
];
